Add Duracloud

Hwesta opened the pull request:
Add Duracloud as a Space.

Note this requires #28 and #31

Hwesta committed:
Duracloud: Stub out Space

refs #7133

Add empty Space, form, constants etc.

Hwesta committed:
Duracloud: Add fields

refs #7133

Add fields to Duracloud Space. Update places to match: constant, forms,
migration, fixtures, test.

Hwesta committed:
Duracloud: Add helpers

refs #7133

Add session with auth for use with all the other functions and
constructor for duraspace URL.

Hwesta committed:
Duracloud: Add move from SS

refs #7133

Hwesta committed:
Duracloud: Add move to SS

refs #7133

Hwesta committed:
Duracloud: Move to SS supports globbing

refs #7133

Hwesta committed:
Duracloud: Add browse

refs #7133

Hwesta committed:
Duracloud: Add delete

refs #7133

Hwesta commented:
We now have two 0005_foo migrations. (South handles this correctly by running them in alphabetical order).  

Previously, we named migrations after the release in question, but we had some problems with that in dev/testing environments. Do we want to switch to a migration per feature or continue updating the release-named migration?

I can rename 0005_v0_5_0.py to 0005_fedora.py and 0005_duracloud.py to 0006_duracloud.py if we go the per-feature route.

mistydemeo commented:
> I can rename 0005_v0_5_0.py to 0005_fedora.py and 0005_duracloud.py to 0006_duracloud.py if we go the per-feature route.

I would really, really like to do this! It makes it much easier for me to keep the storage service around for extended periods, running smaller migrations forward and backwards as necessary.

mistydemeo commented on storage_service/locations/models/duracloud.py:
```diff
@@ -0,0 +1,138 @@
+        paths = self._get_files_list(path)
+        paths = [p.replace(path, '', 1) for p in paths]
+        for p in paths:
+            print p
```
Debugging?

Hwesta replied:
Whoops, yes. Removed.

mistydemeo commented on storage_service/locations/models/duracloud.py:
```diff
@@ -0,0 +1,138 @@
+        for p in paths:
+            print p
+        entries = sorted(list(set(p.split('/')[0] for p in paths)))
+        directories = sorted(list(set(p.split('/')[0] for p in paths if len(p.split('/')) > 1)))
```
You can run `sorted()` on a `set` directly without needing to create an in-memory list first.

Hwesta replied:
Nice catch. Extra list removed.

Hwesta commented:
`0005_duracloud.py` renamed `0006_duracloud.py`.  Renaming of the fedora/0_5_0 migration done in #34

Hwesta committed:
Duracloud: Space path is optional

refs #7133

Hwesta committed:
Duracloud: Browse works without leading /

refs #7133

Browse resources should only join with the Space/Location path if it does
not contain the Space/Location path already.  Some calls to browse contain
the full path, and others relative paths.  This should work for Spaces with
or without a leading / in the path.

Update directory browser JS in SS to handle paths that don't begin with / as
well.

Hwesta committed:
Duracloud: Minor fixes

Don't display password, show the space path as <None> if it doesn't
exist.

mistydemeo commented on requirements/base.txt:
```diff
@@ -7,9 +7,8 @@ South==0.8.4
 bagit==1.3.7
 django-annoying==0.7.7
 git+https://github.com/swordapp/python-client-sword2.git@0c241438f2ced13ec3364caa03edf28d5ed5c02c
+requests>=2.3.0
```
Add a newline to the end of the file?

mistydemeo commented on requirements/local.txt:
```diff
@@ -7,3 +7,4 @@ Sphinx==1.2b1
 ipython==1.1.0
 pytest
 pytest-django
+vcrpy>=1.0.0
```
Ditto

Hwesta replied:
Done for both.

mistydemeo commented on storage_service/locations/models/duracloud.py:
```diff
@@ -0,0 +1,144 @@
+
+    @property
+    def duraspace_url(self):
+        return 'https://' + self.host + '/durastore/' + self.duraspace + '/'
```
Use `urlparse.urljoin` for this? https://docs.python.org/2/library/urlparse.html#urlparse.urljoin

Hwesta replied:
I don't think this is worthwhile. Unlike `os.path.join`,  `urlparse.urljoin` only accepts two parameters - the base and relative path.  I would have to make almost all the URL with `+` before passing it into urljoin, which seems silly at that point.  eg `urlparse.urljoin('https://' + self.host, 'durastore/' + self.duraspace + '/')`

mistydemeo replied:
Oh, good point - makes sense. Never mind, in that case.

mistydemeo commented on storage_service/locations/models/duracloud.py:
```diff
@@ -0,0 +1,144 @@
+        url = self.duraspace_url + delete_path
+        response = self.session.delete(url)
+        if response.status_code == 404:
+            # File cannot be found - this may be a folder
```
We can't check ahead of time if it's a folder?

Hwesta replied:
Duracloud is object storage - folders don't actually exist. To make it play nice with the existing structure and Archivematica I have it fake folders by looking for the provided path as prefix when it can't find something. If it still doesn't find anything, then the 'folder' must not exist.

mistydemeo replied:
Ah, okay.

mistydemeo commented on storage_service/locations/models/duracloud.py:
```diff
@@ -0,0 +1,144 @@
+            url = self.duraspace_url + destination_path
+            self._upload_file(url, source_path)
+        else:
+            raise StorageException('%s is neither a file nor a directory, may not exist' % source_path)
```
Ass an explicit `os.path.exists` check?

Hwesta replied:
Done

mistydemeo commented on storage_service/locations/tests/test_duracloud.py:
```diff
@@ -0,0 +1,113 @@
+        self.ds_object = models.Duracloud.objects.all()[0]
+
+    def test_has_required_attributes(self):
+        assert self.ds_object.host
```
This is really just testing the test fixture, not the actual code, isn't it?

Hwesta replied:
Yeah. I set it up when I was testing the fixtures and didn't see a reason to remove it.

mistydemeo replied:
Makes sense, just checking I wasn't missing something.

mistydemeo commented on storage_service/locations/tests/test_duracloud.py:
```diff
@@ -0,0 +1,113 @@
+        # Create test.txt
+        open('test.txt', 'w').write('test file\n')
+        # Upload
+        self.ds_object.move_from_storage_service('test.txt', '/ts/test.txt')
```
Not necessarily key at this stage, but might be nice to mock the filesystem code so we don't have to actually hit the filesystem during the test.

Hwesta replied:
I agree, but I don't know how to do that.

mistydemeo commented on storage_service/locations/tests/test_duracloud.py:
```diff
@@ -0,0 +1,113 @@
+        # Cleanup
+        os.remove('test.txt')
+        auth = requests.auth.HTTPBasicAuth(self.ds_object.user, self.ds_object.password)
+        requests.delete('https://' + self.ds_object.host + '/durastore/' + self.ds_object.duraspace + '/ts/test.txt', auth=auth)
```
`urljoin` here too?

Hwesta replied:
Probably not, for reasons stated above.

mistydemeo commented on storage_service/locations/tests/test_duracloud.py:
```diff
@@ -0,0 +1,113 @@
+        # Cleanup
+        os.remove('folder/test/test.txt')
+        os.remove('folder/test/subfolder/test2.txt')
+        # BUG Globbing may match and fetch extra things (because of use of normpath)
```
Hm, I'm a bit nervous about this. Is this out of scope to fix right now?
